File: python/flink_agents/api/skills/skill_parser.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
import re
from dataclasses import dataclass
from typing import Dict

import yaml

from flink_agents.api.skills.agent_skill import AgentSkill


@dataclass
class ParsedMarkdown:
    """Result of parsing markdown with frontmatter.

    Contains both the extracted metadata and the markdown content.

    Attributes:
    ----------
    metadata : Dict[str, str]
        YAML metadata extracted from frontmatter.
    content : str
        Markdown content (without frontmatter).
    """

    metadata: Dict[str, str]
    content: str


class MarkdownSkillParser:
    """Utility for parsing Markdown files with YAML frontmatter."""

    # Pattern to match frontmatter: starts with ---, ends with ---
    # Group 1: frontmatter content (non-greedy)
    # Group 2: remaining content
    FRONTMATTER_PATTERN = re.compile(
        r"^---\s*[\r\n]+(.*?)[\r\n]*---(?:\s*[\r\n]+)?(.*)", re.DOTALL
    )

    # Pattern to match key: value format
    KEY_VALUE_PATTERN = re.compile(r"^([a-zA-Z_][a-zA-Z0-9_-]*)\s*:\s*(.*)$")

    @classmethod
    def parse(cls, markdown: str) -> ParsedMarkdown:
        """Parse markdown content with YAML frontmatter.

        Extracts both the YAML metadata and the markdown content.

        Args:
            markdown: Markdown content.

        Returns:
            ParsedMarkdown containing metadata and content.

        Raises:
            ValueError: If YAML syntax is invalid.
        """
        if not markdown:
            return ParsedMarkdown(metadata={}, content="")

        matcher = cls.FRONTMATTER_PATTERN.match(markdown)

        if not matcher:
            # No frontmatter found, treat entire content as markdown
            return ParsedMarkdown(metadata={}, content=markdown)

        yaml_content = matcher.group(1).strip()
        markdown_content = matcher.group(2)

        if not yaml_content:
            return ParsedMarkdown(metadata={}, content=markdown_content)

        try:
            metadata = yaml.safe_load(yaml_content)
            return ParsedMarkdown(metadata=metadata, content=markdown_content)
        except ValueError:
            raise
        except Exception as e:
            raise ValueError(f"Invalid YAML frontmatter syntax: {e}") from e


class SkillParser:
    """Parser for creating AgentSkill from SKILL.md content.

    This class provides high-level parsing functionality to create
    AgentSkill instances from raw SKILL.md file content and resources.
    """

    @classmethod
    def parse_skill(
        cls,
        skill_md_content: str,
    ) -> "AgentSkill":
        """Parse SKILL.md content and create an AgentSkill.

        Args:
            skill_md_content: The raw content of SKILL.md file.

        Returns:
            The parsed AgentSkill.

        Raises:
            ValueError: If required fields (name, description) are missing.
        """
        parsed = MarkdownSkillParser.parse(skill_md_content)
        metadata = parsed.metadata

        # Required fields
        name = metadata.get("name")
        description = metadata.get("description")

        if not name:
            raise ValueError("The SKILL.md must have a YAML frontmatter including 'name' field.")
        if not description:
            raise ValueError("The SKILL.md must have a YAML frontmatter including 'description' field.")
        
        if not parsed.content:
            raise ValueError("The SKILL.md must have a markdown content after YAML frontmatter.")

        return AgentSkill(
            name=name.strip(),
            description=description.strip(),
            license=metadata.get("license"),
            compatibility=metadata.get("compatibility"),
            metadata=metadata.get("metadata"),
            content=parsed.content,
        )

File: runtime/src/main/java/org/apache/flink/agents/runtime/skill/SkillParser.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.runtime.skill;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Parser that splits a {@code SKILL.md} file into YAML frontmatter and markdown body, then
 * constructs an {@link AgentSkill}.
 *
 * <p>Mirrors the Python {@code flink_agents.runtime.skill.skill_parser}.
 */
public final class SkillParser {

    private static final Pattern FRONTMATTER =
            Pattern.compile(
                    "^---\\s*[\\r\\n]+(.*?)[\\r\\n]*---(?:\\s*[\\r\\n]+)?(.*)", Pattern.DOTALL);

    private static final ObjectMapper YAML = new ObjectMapper(new YAMLFactory());

    private SkillParser() {}

    /** Result of splitting a markdown file. */
    public static final class ParsedMarkdown {
        public final Map<String, Object> metadata;
        public final String content;

        public ParsedMarkdown(Map<String, Object> metadata, String content) {
            this.metadata = metadata == null ? Collections.emptyMap() : metadata;
            this.content = content == null ? "" : content;
        }
    }

    /** Split the markdown into YAML frontmatter and the remaining body. */
    public static ParsedMarkdown parseMarkdown(String markdown) {
        if (markdown == null || markdown.isEmpty()) {
            return new ParsedMarkdown(Collections.emptyMap(), "");
        }
        Matcher m = FRONTMATTER.matcher(markdown);
        if (!m.matches()) {
            return new ParsedMarkdown(Collections.emptyMap(), markdown);
        }
        String yaml = m.group(1).trim();
        String body = m.group(2);
        if (yaml.isEmpty()) {
            return new ParsedMarkdown(Collections.emptyMap(), body);
        }
        try {
            Map<String, Object> metadata =
                    YAML.readValue(yaml, new TypeReference<Map<String, Object>>() {});
            return new ParsedMarkdown(metadata, body);
        } catch (Exception e) {
            throw new IllegalArgumentException(
                    "Invalid YAML frontmatter syntax: " + e.getMessage(), e);
        }
    }

    /** Parse a SKILL.md content into an {@link AgentSkill}. */
    public static AgentSkill parseSkill(String skillMdContent) {
        ParsedMarkdown parsed = parseMarkdown(skillMdContent);
        Map<String, Object> metadata = parsed.metadata;

        Object name = metadata.get("name");
        if (!(name instanceof String) || ((String) name).trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "The SKILL.md must have a YAML frontmatter including 'name' field.");
        }
        Object description = metadata.get("description");
        if (!(description instanceof String) || ((String) description).trim().isEmpty()) {
            throw new IllegalArgumentException(
                    "The SKILL.md must have a YAML frontmatter including 'description' field.");
        }
        if (parsed.content == null || parsed.content.isEmpty()) {
            throw new IllegalArgumentException(
                    "The SKILL.md must have a markdown content after YAML frontmatter.");
        }

        Object license = metadata.get("license");
        Object compatibility = metadata.get("compatibility");
        Object inner = metadata.get("metadata");
        Map<String, String> innerMetadata = null;
        if (inner instanceof Map) {
            innerMetadata = new HashMap<>();
            for (Map.Entry<?, ?> e : ((Map<?, ?>) inner).entrySet()) {
                innerMetadata.put(String.valueOf(e.getKey()), String.valueOf(e.getValue()));
            }
        }

        return new AgentSkill(
                ((String) name).trim(),
                ((String) description).trim(),
                parsed.content,
                license == null ? null : license.toString(),
                compatibility == null ? null : compatibility.toString(),
                innerMetadata);
    }
}
